// mindee/src/v2/http/mindeeApiV2.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _MindeeApiV2_instances, _MindeeApiV2_processResponse, _MindeeApiV2_reqPostProductEnqueue, _MindeeApiV2_reqGetJob, _MindeeApiV2_reqGetProductResult;
import { ApiSettings } from "./apiSettings.js";
import { ErrorResponse, JobResponse, } from "../../v2/parsing/index.js";
import { sendRequestAndReadResponse } from "../../http/apiCore.js";
import { LocalInputSource } from "../../input/index.js";
import { MindeeDeserializationError, MindeeError } from "../../errors/index.js";
import { MindeeHttpErrorV2 } from "./errors.js";
import { logger } from "../../logger.js";
/**
 * Mindee V2 API handler.
 */
export class MindeeApiV2 {
    constructor(dispatcher, apiKey) {
        _MindeeApiV2_instances.add(this);
        this.settings = new ApiSettings({ dispatcher: dispatcher, apiKey: apiKey });
    }
    /**
     * Sends a file to the product inference queue.
     * @param product product to enqueue.
     * @param inputSource Local file loaded as an input.
     * @param params {ExtractionParameters} parameters relating to the enqueueing options.
     * @throws Error if the server's response contains an error.
     * @returns a `Promise` containing a job response.
     */
    async enqueueProduct(product, inputSource, params) {
        await inputSource.init();
        const result = await __classPrivateFieldGet(this, _MindeeApiV2_instances, "m", _MindeeApiV2_reqPostProductEnqueue).call(this, product, inputSource, params);
        if (result.data.error !== undefined) {
            throw new MindeeHttpErrorV2(result.data.error);
        }
        return __classPrivateFieldGet(this, _MindeeApiV2_instances, "m", _MindeeApiV2_processResponse).call(this, result, JobResponse);
    }
    /**
     * Get the specified Job.
     * Throws an error if the server's response contains an error.
     * @param jobId The Job ID as returned by the enqueue request.
     * @returns a `Promise` containing the job response.
     */
    async getJob(jobId) {
        const response = await __classPrivateFieldGet(this, _MindeeApiV2_instances, "m", _MindeeApiV2_reqGetJob).call(this, jobId);
        return __classPrivateFieldGet(this, _MindeeApiV2_instances, "m", _MindeeApiV2_processResponse).call(this, response, JobResponse);
    }
    /**
     * Get the result of a queued document from the API.
     * Throws an error if the server's response contains an error.
     * @param product
     * @param inferenceId The inference ID for the result.
     * @returns a `Promise` containing the parsed result.
     */
    async getProductResultById(product, inferenceId) {
        const queueResponse = await __classPrivateFieldGet(this, _MindeeApiV2_instances, "m", _MindeeApiV2_reqGetProductResult).call(this, `https://${this.settings.hostname}/v2/products/${product.slug}/results/${inferenceId}`);
        return __classPrivateFieldGet(this, _MindeeApiV2_instances, "m", _MindeeApiV2_processResponse).call(this, queueResponse, product.responseClass);
    }
    /**
     * Get the result of a queued document from the API.
     * Throws an error if the server's response contains an error.
     * @param product
     * @param url The URL as returned by a Job's resultUrl property.
     * @returns a `Promise` containing the parsed result.
     */
    async getProductResultByUrl(product, url) {
        const queueResponse = await __classPrivateFieldGet(this, _MindeeApiV2_instances, "m", _MindeeApiV2_reqGetProductResult).call(this, url);
        return __classPrivateFieldGet(this, _MindeeApiV2_instances, "m", _MindeeApiV2_processResponse).call(this, queueResponse, product.responseClass);
    }
}
_MindeeApiV2_instances = new WeakSet(), _MindeeApiV2_processResponse = function _MindeeApiV2_processResponse(result, responseClass) {
    if (result.messageObj?.statusCode
        && (result.messageObj?.statusCode > 399 || result.messageObj?.statusCode < 200)) {
        if (result.data?.status !== null) {
            throw new MindeeHttpErrorV2(new ErrorResponse(result.data));
        }
        throw new MindeeHttpErrorV2(new ErrorResponse({
            status: result.messageObj?.statusCode ?? -1,
            title: "Unknown Error",
            detail: result.data?.detail ?? "The server returned an Unknown error.",
            code: `${result.messageObj?.statusCode ?? -1}-000`,
        }));
    }
    try {
        return new responseClass(result.data);
    }
    catch (e) {
        logger.error(`Raised '${e}' Couldn't deserialize response object:\n${JSON.stringify(result.data)}`);
        throw new MindeeDeserializationError("Couldn't deserialize response object.");
    }
}, _MindeeApiV2_reqPostProductEnqueue = 
/**
 * Sends a document to the inference queue.
 * @param product Product to enqueue.
 * @param inputSource Local or remote file as an input.
 * @param params {ExtractionParameters} parameters relating to the enqueueing options.
 */
async function _MindeeApiV2_reqPostProductEnqueue(product, inputSource, params) {
    const form = params.getFormData();
    if (inputSource instanceof LocalInputSource) {
        form.set("file", new Blob([inputSource.fileObject]), inputSource.filename);
    }
    else {
        form.set("url", inputSource.url);
    }
    const path = `/v2/products/${product.slug}/enqueue`;
    const options = {
        method: "POST",
        headers: this.settings.baseHeaders,
        hostname: this.settings.hostname,
        path: path,
        body: form,
        timeoutSecs: this.settings.timeoutSecs,
    };
    return await sendRequestAndReadResponse(this.settings.dispatcher, options);
}, _MindeeApiV2_reqGetJob = async function _MindeeApiV2_reqGetJob(jobId) {
    const options = {
        method: "GET",
        headers: this.settings.baseHeaders,
        hostname: this.settings.hostname,
        path: `/v2/jobs/${jobId}`,
        timeoutSecs: this.settings.timeoutSecs,
    };
    return await sendRequestAndReadResponse(this.settings.dispatcher, options);
}, _MindeeApiV2_reqGetProductResult = 
/**
 * Make a request to GET the status of a document in the queue.
 * @param url URL path to the result.
 * @returns a `Promise` containing the parsed result.
 */
async function _MindeeApiV2_reqGetProductResult(url) {
    const options = {
        method: "GET",
        headers: this.settings.baseHeaders,
        timeoutSecs: this.settings.timeoutSecs,
    };
    if (!url.startsWith("https://")) {
        throw new MindeeError(`Invalid URL: ${url}`);
    }
    return await sendRequestAndReadResponse(this.settings.dispatcher, options, url);
};
