// mindee/src/v1/product/barcodeReader/barcodeReaderV1Document.js
import { cleanOutString, } from "../../../v1/parsing/common/index.js";
import { StringField } from "../../../v1/parsing/standard/index.js";
/**
 * Barcode Reader API version 1.0 document data.
 */
export class BarcodeReaderV1Document {
    constructor(rawPrediction, pageId) {
        /** List of decoded 1D barcodes. */
        this.codes1D = [];
        /** List of decoded 2D barcodes. */
        this.codes2D = [];
        if (rawPrediction["codes_1d"]) {
            rawPrediction["codes_1d"].map((itemPrediction) => this.codes1D.push(new StringField({
                prediction: itemPrediction,
                pageId: pageId,
            })));
        }
        if (rawPrediction["codes_2d"]) {
            rawPrediction["codes_2d"].map((itemPrediction) => this.codes2D.push(new StringField({
                prediction: itemPrediction,
                pageId: pageId,
            })));
        }
    }
    /**
     * Default string representation.
     */
    toString() {
        const codes1D = this.codes1D.join("\n              ");
        const codes2D = this.codes2D.join("\n              ");
        const outStr = `:Barcodes 1D: ${codes1D}
:Barcodes 2D: ${codes2D}`.trimEnd();
        return cleanOutString(outStr);
    }
}
