// mindee/src/v1/product/us/bankCheck/bankCheckV1Page.js
import { cleanOutString } from "../../../../v1/parsing/common/index.js";
import { PositionField } from "../../../../v1/parsing/standard/index.js";
import { BankCheckV1Document } from "./bankCheckV1Document.js";
/**
 * Bank Check API version 1.1 page data.
 */
export class BankCheckV1Page extends BankCheckV1Document {
    constructor(rawPrediction, pageId) {
        super(rawPrediction, pageId);
        /** List of signature positions */
        this.signaturesPositions = [];
        this.checkPosition = new PositionField({
            prediction: rawPrediction["check_position"],
            pageId: pageId,
        });
        rawPrediction["signatures_positions"].map((itemPrediction) => this.signaturesPositions.push(new PositionField({
            prediction: itemPrediction,
            pageId: pageId,
        })));
    }
    toString() {
        const signaturesPositions = this.signaturesPositions.join("\n                      ");
        let outStr = `:Check Position: ${this.checkPosition}
:Signature Positions: ${signaturesPositions}`.trimEnd();
        outStr += "\n" + super.toString();
        return cleanOutString(outStr);
    }
}
