// mindee/src/pdf/pdfUtils.js
import { MindeePdfError } from "../errors/index.js";
import { loadOptionalDependency } from "../dependency/index.js";
function getConcatenatedText(pages) {
    return pages.flatMap(page => page.content.map(item => item.str)).join(" ");
}
/**
 * Extracts text from a full PDF document.
 *
 * @returns A Promise containing the extracted text as a string.
 * @param pdfBuffer PDF handle, as a buffer.
 */
export async function extractTextFromPdf(pdfBuffer) {
    const pdfJsExtract = await loadOptionalDependency("pdf.js-extract", "PDF text extraction");
    const pdfJs = pdfJsExtract.pdfjs || pdfJsExtract;
    const pdfExtract = new pdfJs.PDFExtract();
    const options = {};
    const pdf = await new Promise((resolve, reject) => {
        pdfExtract.extractBuffer(pdfBuffer, options, (err, result) => {
            if (err)
                reject(err);
            if (result === undefined)
                reject(new MindeePdfError("Couldn't process result."));
            else
                resolve(result);
        });
    });
    const pages = pdf.pages.map((page, index) => ({
        pageNumber: index + 1,
        content: page.content.map((item) => ({
            str: item.str,
            x: item.x,
            y: item.y,
            width: item.width,
            height: item.height,
            fontName: item.fontName,
        })),
    }));
    return {
        pages,
        getConcatenatedText: () => getConcatenatedText(pages),
    };
}
/**
 * Checks if a PDF contains source text.
 *
 * @param pdfData Buffer representing the content of the PDF file.
 *
 * @returns A Promise containing a boolean indicating if the PDF has source text.
 */
export async function hasSourceText(pdfData) {
    const text = await extractTextFromPdf(pdfData);
    return text.getConcatenatedText().trim().length > 0;
}
